File: January2022/Python/minimum_number_of_arrows_to_burst_balloons.py
# Author :  Karthik K (https://www.linkedin.com/in/karthikmunirathinam/)

# TC : O(nlogn)
# SC : O(1)

class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        points.sort(key=lambda x: x[0])
        arrowCount = 0
        x = points[0][1]
        for i in range(1, len(points)):
            if points[i][0] <= x:
                x = min(points[i][1], x)
            else:
                arrowCount += 1
                x = points[i][1]
        return arrowCount+1

File: Leetcode2023/January2023/Java/MinimumNumberOfArrowsToBurstBalloons.java
// @saorav21994

class Solution {
    public int findMinArrowShots(int[][] points) {
        
        Arrays.sort(points, (a,b) -> Integer.compare(a[0], b[0]));
        
        // for (int [] point : points) {
        //     for (int p : point) {
        //         System.out.print(p + " ");
        //     }
        //     System.out.println();
        // }
        
        int cnt = 0;
        int left = points[0][0];
        int right = points[0][1];
        for (int [] point : points) {
            if (point[0] <= right) {
                left = point[0];
                right = Math.min(right, point[1]);
            }
            else {
                cnt += 1;
                left = point[0];
                right = point[1];
            }
        }
        
        return cnt+1;
    }
}
